specifyweb/frontend/js_src/lib/components/Formatters/Formatter.tsx

```diff
@@ -74,40 +102,64 @@ function Definitions({
   readonly item: GetSet<Formatter>;
 }): JSX.Element | null {
   const isReadOnly = React.useContext(ReadOnlyContext);
-  const handleChange = (fields: Formatter['definition']['fields']): void =>
-    setFormatter({
-      ...formatter,
-      definition: {
-        ...formatter.definition,
-        fields,
-      },
-    });
+
+  const handleChange = React.useCallback(
+    (fields: Formatter['definition']['fields']): void => {
+      setFormatter({
+        ...formatter,
+        definition: {
+          ...formatter.definition,
+          fields,
+        },
+      });
+    },
+    [formatter, setFormatter]
+  );
+
   const hasCondition = formatter.definition.conditionField !== undefined;
   const trimmedFields = hasCondition
     ? formatter.definition.fields
     : formatter.definition.fields.slice(0, 1);
 
   const table = formatter.table;
+
+  const handleChanged = (
+    field: Formatter['definition']['fields'][number],
+    index: number
+  ): void =>
+    handleChange(replaceItem(formatter.definition.fields, index, field));
+
+  React.useEffect(() => {
+    if (formatter.definition.fields.length === 0) {
+      handleChange([
+        ...formatter.definition.fields,
+        {
+          value: undefined,
+          fields: [],
+        },
+      ]);
+    }
+  }, [formatter.definition.fields, handleChange]);
+
   return table === undefined ? null : (
     <div className="flex flex-col gap-4 divide-y divide-gray-500 [&>*]:pt-4">
-      {trimmedFields.map(({ value, fields }, index) => {
-        const handleChanged = (
-          field: Formatter['definition']['fields'][number]
-        ): void =>
-          handleChange(replaceItem(formatter.definition.fields, index, field));
```

Comment:
(optional)
while it might have looked a bit weird to define the function inside a .map(), I feel like the code was simpler that way because you didn't need to explicitly pass the index argument.

defining functions inside of .map() shouldn't feel weird because you do that all the time (the first argument to .map() it an arrow function too after all), and each callback handler is an arrow function too

Reply:
it was to access it in my use effect